src/terms/local.jl

```diff
@@ -136,28 +153,33 @@ function compute_forces(::TermAtomicLocal, basis::PlaneWaveBasis{T}, ψ, occupat
 end
 @timing "forces: local" function forces_local(S, basis::PlaneWaveBasis{T}, ρ, q) where {T}
     model = basis.model
-    recip_lattice = model.recip_lattice
-    ρ_fourier = to_cpu(fft(basis, total_density(ρ)))
     real_ifSreal = S <: Real ? real : identity
 
-    # TODO: currently, local forces entierly computed on the CPU.
-    #       This might not be optimal. See compute_local_potential() too.
     Gqs_cart = [model.recip_lattice * (G + q) for G in to_cpu(G_vectors(basis))]
-    form_factors = atomic_local_form_factors(basis, Gqs_cart)
+    form_factors, iG2norm = atomic_local_form_factors(basis, Gqs_cart)
+
+    Gqs = map(G -> G+q, G_vectors(basis))
+    ρ_fourier = reshape(fft(basis, total_density(ρ)), length(Gqs))
+
+    indices = to_device(basis.architecture, collect(1:length(Gqs)))
+    ρpot = similar(ρ_fourier)
+    twoπp = similar(ρ_fourier)
 
     # energy = sum of form_factor(G) * struct_factor(G) * rho(G)
     # where struct_factor(G) = e^{-i G·r}
     forces = [zero(Vec3{S}) for _ = 1:length(model.positions)]
     for (igroup, group) in enumerate(model.atom_groups)
-        element = model.atoms[first(group)]
         for idx in group
             r = model.positions[idx]
-            forces[idx] = -real_ifSreal(sum(conj(ρ_fourier[iG])
-                                              * form_factors[(igroup, norm(Gqs_cart[iG]))]
-                                              * cis2pi(-dot(G + q, r))
-                                              * (-2T(π)) * (G + q) * im
-                                              / sqrt(model.unit_cell_volume)
-                                        for (iG, G) in enumerate(to_cpu(G_vectors(basis)))))
+
+            ff_group = @view form_factors[:, igroup]
+            map!(iG -> cis2pi(-dot(Gqs[iG], r))*conj(ρ_fourier[iG])*
+                      ff_group[iG2norm[iG]], ρpot, indices)
+
+            forces[idx] += map(1:3) do α
+                map!(p -> -2π*im*p[α], twoπp, Gqs)
+                -real_ifSreal(sum(ρpot .* twoπp)/sqrt(model.unit_cell_volume))
```

Comment:
Should we not go all the way and also avoid the allocation of `twoπρ and the temporary in `ρpot .* twoπρ` by dirrectly doing the sum ?